### ISSUE TITLE
Assertion failure in steal_into / steal_into2

### ISSUE THREAD

**Issue by jebrosen** [tokio-rs/tokio#2382]
## Description

A "hello_world" using Rocket's async branch is hitting an assertion in `queue.rs`.

## Reproduction

`Cargo.toml`:

```toml
[dependencies]
rocket = { git = "https://github.com/SergioBenitez/Rocket.git", branch = "async" } # As of 2020-04-06 this is rev="78c8ac8"
tokio = "=0.2.16"

[profile.release]
lto = true
codegen-units = 1
```

`main.rs`:

```rust
#![feature(proc_macro_hygiene, decl_macro)]

use rocket::{get, routes};

#[get("/")]
fn hello() -> String {
    "Hello world!".to_string()
}

fn main() {
    rocket::ignite().mount("/", routes![hello]).launch().unwrap();
}
```

Then run `wrk -t20 -c1000 -d1h http://localhost:8000` until it fails.

The following assertion is failing in `queue.rs` in both 0.2.15 and 0.2.16:

```
thread 'tokio-runtime-worker' panicked at 'assertion failed: `(left != right)`
  left: `117`,
 right: `117`', /home/me/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.16/src/runtime/queue.rs:414:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

~~In 0.2.14, (EDIT: I misread something in the code; this was nothing more than the already-fixed issue #2362)~~

## Speculation

So far it has not been been reproduced on 0.2.13, suggesting that 0.2.14 includes the regressed commit. My first guess was that it had to do with the scheduler refactor #2273, but after reviewing the change log I also suspect #2315 since it specifically deals with work-stealing. I will try to bisect to a specific commit, but it is a slow process because it takes some time before the assertion fails.

---

This issue was discovered by @ThouCheese, and we did the initial triage in rocket's IRC channel. All results were collected on Linux on an Intel i7-8750H. I have also been able to reproduce the assertion failure on a few commits, on Linux on an AMD Ryzen 7 3700U.

**Comment by jebrosen** [comment]
#2349 (caa7e180) is another possible culprit. I have verified that it is *a* bad commit, but I don't yet know if it is the first bad commit.

**Comment by carllerche** [comment]
Thanks for the report... i'm not sure what is going on yet. It would be helpful to try to get a more minimal repro. I am starting to dig in.

**Comment by carllerche** [comment]
I added more loom tests [here](https://github.com/tokio-rs/tokio/commit/d2f7d15ab4e2256328ad29415961ba1735979429) but was not able to trigger this assertion.

**Comment by llitz** [comment]
Had this error happen a couple times with a program processing around 100MB of data per second. I am not using rocket, just processing kafka events.

The only thing we have in common is tokio and lto=true

0.2.13 is indeed stable for me (has been for a while actually)

**Comment by ThouCheese** [comment]
@llitz I was also able to reproduce this without lto enabled.

**Comment by carllerche** [comment]
I ran the repro for about ~10 minutes without being able to repro. I added a stress test here: https://github.com/tokio-rs/tokio/blob/queue-maybe-bug/tokio/src/runtime/tests/queue.rs#L32-L88 which did not trigger it either.

I will try to add some more assertions to try to identify the issue earlier.

**Comment by carllerche** [comment]
There are two options for the cause of the bug:

* Invalid logic in `queue.rs` related to the atomic ops
* Memory corruption causes by something else.

I'm going to assume the first option for now.

The next question is why do the loom tests not trigger the assertion. As far as I can tell, the options are:

* No loom test covers that case
* It is related to having a queue capacity of 256 (loom tests limit the queue capacity to 4).
* There is a bug in loom.

**Comment by llitz** [comment]
@carllerche 
For me it only triggers after running my program for a very long time, not necessarily under heavy load (it triggered when processing around 60MB/s too).
My program is running on EC2 AWS Servers. The atomic seems like a possibility.

One other point I observed is that with 0.2.16 my program behaves differently...
Basically I spawn 4 jobs that go through a loop. The jobs each run their own loops 1,000,000 times, wait for data to be sent and restart the loop. 
On 0.2.16, for some weird reason, the loops would only restart after all 4 jobs run through each loop. On 0.2.13, they run completely independent of each other.

I don't know the internals of Tokio and I am unsure if this is relevant for our left != right

**Comment by carllerche** [comment]
@llitz there was a recent scheduler change: https://tokio.rs/blog/2020-04-preemption/

**Comment by carllerche** [comment]
I have a PR to fix the bug (#2387). @jebrosen has already validated that it fixes his issue. If anyone else is able to check, it would be appreciated.
